### PR TITLE
Improve error messages for spice ai connector, separate errors to different lines for DuckDB, Delta Lake, Databricks connector

### PR DESCRIPTION
**Pull request by Sevenannn**
## 🗣 Description

Improve error messages for spice ai connectors, including changes in spiceai, flight client, and flight data component, following the principles defined in https://github.com/spiceai/spiceai/blob/trunk/docs/dev/error_handling.md

This PR also include changes to separate errors including sources to separate lines for DuckDB, Delta Lake, Databricks connector.

### CODE REVIEW

File: crates/flight_client/src/lib.rs [hunk 1]
```diff
@@ -42,40 +42,46 @@ pub mod tls;
...
 
-    #[snafu(display("Invalid metadata value: {source}"))]
+    #[snafu(display("Authentication failed: {source}\nEnsure the credentials are valid."))]
     InvalidMetadata {
         source: tonic::metadata::errors::InvalidMetadataValue,
     },
 
-    #[snafu(display("Unable to perform handshake: {source}"))]
+    #[snafu(display(
+        "Unable to perform handshake with flight endpoint: {source}\nVerify the configuration and try again."
```
Comment (sgrebnov) [inline crates/flight_client/src/lib.rs]:
```suggestion
        "Failed to complete the handshake with the flight endpoint: {source}\nVerify authentication settings and try again."
```

Comment (peasee) [reply crates/flight_client/src/lib.rs]: Is a handshake something a user cares about? Maybe stating the connection failed would be clearer:

```suggestion
        "Failed to connect to the Flight server: {source}\nVerify the configuration and try again."
```

Comment (sgrebnov) [reply crates/flight_client/src/lib.rs]: I like the more user-friendly version as well. We just need to make sure we are not using the same error message so we can distinguish between use cases where users encounter and report the error. The intent was to add something to `Failed to connect to the Flight server` so we understand where it is happening

Comment (Sevenannn) [reply crates/flight_client/src/lib.rs]: I updated this error to `Failed to connect to the Flight server: Handshake failed.` To provide a user-friendly error, as well as retaining a minimal information about handshake, IMO handshake context information could be helpful for user, as well as our internal debugging purpose.

**Review by sgrebnov — DISMISSED** [review]

File: crates/flight_client/src/lib.rs [hunk 2]
```diff
@@ -118,40 +118,48 @@ impl std::fmt::Display for TonicStatusMessage {
...
 
-    #[snafu(display("Failed to execute query. Invalid metadata value.\n{source}"))]
+    #[snafu(display("Authentication failed.\n{source}\nEnsure the credentials are valid."))]
     InvalidMetadata {
         source: tonic::metadata::errors::InvalidMetadataValue,
     },
 
-    #[snafu(display("Failed to connect to the Flight server.\n{source}"))]
+    #[snafu(display(
+        "Failed to connect to Flight server: Handshake failed.\n{source}\nVerify the configuration and try again."
```
Comment (peasee) [inline crates/flight_client/src/lib.rs]:
```suggestion
        "Failed to connect to Flight server: Handshake failed.\n{source}"
```

The new tonic status messages print their own actions, so let's not repeat an action here.

File: crates/flight_client/src/lib.rs [hunk 3]
```diff
@@ -118,40 +118,48 @@ impl std::fmt::Display for TonicStatusMessage {
...
     UnableToConvertSchema { source: arrow::error::ArrowError },
 
-    #[snafu(display("Failed to execute query.\n{source}"))]
+    #[snafu(display("Query execution failed.\n{source}"))]
     UnableToQuery {
         source: Box<dyn std::error::Error + Send + Sync>,
     },
 
-    #[snafu(display("Failed to write data.\n{source}"))]
+    #[snafu(display("Failed to publish data to flight endpoint.\n{source}\nVerify the configuration and try again"))]
```
Comment (peasee) [inline crates/flight_client/src/lib.rs]:
```suggestion
    #[snafu(display("Failed to publish data to flight endpoint.\n{source}"))]
```

File: crates/flight_client/src/lib.rs [hunk 4]
```diff
@@ -118,40 +118,48 @@ impl std::fmt::Display for TonicStatusMessage {
...
     UnableToPublish { source: TonicStatusError },
 
-    #[snafu(display("Failed to execute query. Authorization failed.\nEnsure that the username and password are correctly configured, and have the necessary permissions."))]
+    #[snafu(display("Unauthorized. Verify the credentials."))]
```
Comment (peasee) [inline crates/flight_client/src/lib.rs]:
```suggestion
    #[snafu(display("Unauthorized. Verify the credentials are configured correctly."))]
```

File: crates/runtime/src/dataconnector/databricks.rs [hunk 5]
```diff
@@ -47,7 +47,7 @@ pub enum Error {
     #[snafu(display("Invalid `databricks_use_ssl` value: '{value}'. Use 'true' or 'false'.\nFor details, visit: https://docs.spiceai.org/components/data-connectors/databricks#parameters"))]
     InvalidUsessl { value: String },
 
-    #[snafu(display("Failed to connect to Databricks Spark: {source}\nVerify Databricks Connector configurations and try again\nFor details, visit: https://docs.spiceai.org/components/data-connectors/databricks#parameters"))]
+    #[snafu(display("Failed to connect to Databricks Spark.\n{source}\nVerify Databricks Connector configurations and try again\nFor details, visit: https://docs.spiceai.org/components/data-connectors/databricks#parameters"))]
```
Comment (peasee) [inline crates/runtime/src/dataconnector/databricks.rs]:
```suggestion
    #[snafu(display("Failed to connect to Databricks Spark.\n{source}\nVerify the connector configuration, and try again.\nFor details, visit: https://docs.spiceai.org/components/data-connectors/databricks#parameters"))]
```

File: crates/runtime/src/dataconnector/duckdb.rs [hunk 6]
```diff
@@ -38,7 +38,7 @@ use super::{
 
 #[derive(Debug, Snafu)]
 pub enum Error {
-    #[snafu(display("Unable to create DuckDB connection pool: {source}"))]
+    #[snafu(display("Failed to create DuckDB connection pool.\n{source}"))]
```
Comment (peasee) [inline crates/runtime/src/dataconnector/duckdb.rs]:
```suggestion
    #[snafu(display("Failed to connect to DuckDB.\n{source}\nVerify the database configuration, and try again."))]
```

File: crates/runtime/src/dataconnector/spiceai.rs [hunk 7]
```diff
@@ -57,25 +57,22 @@ use tonic::metadata::MetadataMap;
...
 
-    #[snafu(display(r#"Unable to connect to endpoint "{endpoint}": {source}"#))]
+    #[snafu(display(r#"Failed to connect to SpiceAI endpoint "{endpoint}".\n{source}\nEnsure the endpoint is valid and reachable"#))]
     UnableToVerifyEndpointConnection {
         source: ns_lookup::Error,
         endpoint: String,
     },
 
-    #[snafu(display("Unable to create flight client: {source}"))]
+    #[snafu(display("Failed to create flight client.\n{source}"))]
```
Comment (peasee) [inline crates/runtime/src/dataconnector/spiceai.rs:72]: why would a Flight client fail to create? what can a user do to resolve it?

if this is an internal error, perhaps direct users to file a bug.

File: crates/runtime/src/dataconnector/spiceai.rs [hunk 8]
```diff
@@ -57,25 +57,22 @@ use tonic::metadata::MetadataMap;
...
     UnableToCreateFlightClient { source: flight_client::Error },
 
-    #[snafu(display("Unable to get append stream schema: {source}"))]
+    #[snafu(display("Failed to get append stream schema.\n{source}"))]
```
Comment (peasee) [inline crates/runtime/src/dataconnector/spiceai.rs:75]: why would we fail to get an append stream? what can a user do to resolve it?

if this is an internal error, perhaps direct users to file a bug.

Comment (Sevenannn) [reply crates/runtime/src/dataconnector/spiceai.rs:75]: This error serves as an intermediate layer for propagating flight_client::Error, where we have more detailed error on why failing to get the append stream and detailed actions defined. I'm inclined to keep this intermediate level message as it'll be easier for debugging when customer report the error. Please let me know whether this approach is acceptable

**Review by peasee — DISMISSED** [review]

**Review by peasee — APPROVED** [review]

**Pull request merged**